### PR TITLE
CBL-2582: test: Updated Live query 

### PR DESCRIPTION
**Pull request by jayahariv**
* Moved the previous live-query together with these set of tests. 
* Add more tests
* fix: CBL-2582

### CODE REVIEW

File: Objective-C/Tests/QueryTest+Main.m [hunk 1]
```diff
@@ -1873,5 +1794,280 @@ - (void) validateN1QLQuery: (NSString*)queryString {
...
+            AssertEqual(rows.count, 9u);
+        } else {
+            AssertEqual(rows.count, 10u);
+            CBLDocument* doc = [self.db documentWithID: [rows[0] valueAtIndex: 0]];
+            AssertEqualObjects([doc valueForKey: @"number1"], @(-1));
+            [x fulfill];
+        }
+    }];
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
```
Comment (pasin) [inline Objective-C/Tests/QueryTest+Main.m:1824]: I think it is better to not depend on the delay. You can have two expectations and wait until the first one is fulfilled before creating a new set of docs.

File: Objective-C/Tests/QueryTest+Main.m [hunk 2]
```diff
@@ -1873,5 +1794,280 @@ - (void) validateN1QLQuery: (NSString*)queryString {
...
+        AssertNil(change.error);
+        NSArray<CBLQueryResult*>* rows = [change.results allObjects];
+        if (count == 1) {
+            AssertEqual(rows.count, 9u);
+        } else {
+            XCTFail(@"Unexpected update from LiveQuery");
+        }
+    }];
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)),
```
Comment (pasin) [inline Objective-C/Tests/QueryTest+Main.m:1854]: Same comment as above.

File: Objective-C/Tests/QueryTest+Main.m [hunk 3]
```diff
@@ -1873,5 +1794,280 @@ - (void) validateN1QLQuery: (NSString*)queryString {
...
+            
+            // newly added doc notification
+            AssertEqual(rows.count, 2);
+            AssertEqual([rows[0] integerAtIndex: 0], 3);
+            
+            [first fulfill];
+        }
+    }];
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
```
Comment (pasin) [inline Objective-C/Tests/QueryTest+Main.m:1908]: This seems to be not relevant for the test.

File: Objective-C/Tests/QueryTest+Main.m [hunk 4]
```diff
@@ -1873,5 +1794,280 @@ - (void) validateN1QLQuery: (NSString*)queryString {
...
+        [self createDocNumbered: 3 of: 10];
+    });
+    
+    // wait for first set of notifications
+    [self waitForExpectations: @[first] timeout: 10.0];
+    AssertEqual(count, 2);
+    [q removeChangeListenerWithToken: token1];
```
Comment (pasin) [inline Objective-C/Tests/QueryTest+Main.m:1915]: It's better to not remove the token1 and to make sure that after adding the second listener, the first listener doesn't get the notification.

File: Objective-C/Tests/QueryTest+Main.m [hunk 5]
```diff
@@ -1873,5 +1794,280 @@ - (void) validateN1QLQuery: (NSString*)queryString {
...
+                                  orderBy: @[[CBLQueryOrdering property: @"number1"]]];
+    
+    // first listener
+    XCTestExpectation* x = [self expectationWithDescription: @"live-query-finish"];
+    __block int count = 0;
+    id token = [q addChangeListener: ^(CBLQueryChange* change) {
+        count++;
+        NSArray<CBLQueryResult*>* rows = [change.results allObjects];
+        if (count == 1) {
+            // initial notification about already existing result set!
```
Comment (pasin) [inline Objective-C/Tests/QueryTest+Main.m:1951]: Need to be updated?

File: Objective-C/Tests/QueryTest+Main.m [hunk 6]
```diff
@@ -1873,5 +1794,280 @@ - (void) validateN1QLQuery: (NSString*)queryString {
...
+            
+            // deleted the doc should return empty result set!
+            AssertEqual(rows.count, 0);
+            [x fulfill];
+        }
+    }];
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+        NSError* error = nil;
+        Assert([self.db purgeDocumentWithID: @"doc1" error: &error]);
```
Comment (pasin) [inline Objective-C/Tests/QueryTest+Main.m:1964]: This test can be simplified by just have the query returning empty result from the first notification and nothing else.

File: Objective-C/Tests/QueryTest+Main.m [hunk 7]
```diff
@@ -1873,5 +1794,280 @@ - (void) validateN1QLQuery: (NSString*)queryString {
...
+            for (CBLQueryResult* result in change.results) {
+                num++;
+                Assert([result integerAtIndex: 0] < 10);
+            }
+            AssertEqual(num, 10);
+            [y fulfill];
+        }
+    }];
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
```
Comment (pasin) [inline Objective-C/Tests/QueryTest+Main.m:2028]: Trying not to depending on avoid flaky test.

File: Objective-C/Tests/QueryTest+Main.m [hunk 8]
```diff
@@ -1873,5 +1794,280 @@ - (void) validateN1QLQuery: (NSString*)queryString {
...
+        NSArray<CBLQueryResult*>* rows = [change.results allObjects];
+        if (count == 1) {
+            AssertEqual(rows.count, 9);
+        } else if (count == 2) {
+            AssertEqual(rows.count, 4);
+            [x fulfill];
+        }
+    }];
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
```
Comment (pasin) [inline Objective-C/Tests/QueryTest+Main.m:2061]: Same as above. Trying to avoid depending on delay.

**Review by pasin — CHANGES_REQUESTED** [review]
Any comments to Objc-C also apply to Swift tests.

File: Objective-C/Tests/QueryTest+Main.m (hunk 2 above)
Comment (jayahariv) [reply Objective-C/Tests/QueryTest+Main.m:1854]: This cannot be done as previous. we are waiting for next 2 secs to make sure, there are no changes notified with this change. So it's better to keep it as before.

**Review by pasin — APPROVED** [review]

**Pull request merged**